topsites.py

```diff
@@ -0,0 +1,180 @@
+
+
+class TopSites:
+    '''Used to query top site info from alexa API to Site object'''
+    def __init__(self, access_key, secret_key):
+        self.access_key = access_key
+        self.secret_key = secret_key
+
+    def query(self, begin, begin_priority):
+        '''Query alexa with given range, save to Site'''
+        query_string = self.build_query(begin, 100)
+        toSign = "GET\n" + ATS_SERVICE_HOST + "\n/\n" + query_string
+        signature = self.gen_sign(toSign, self.secret_key)
+        uri = "{base}{query}&{signature}".format(
+            base=ATS_AWS_BASE_URL,
+            query=query_string,
+            signature=urlencode({"Signature": signature}, "UTF-8"))
+
+        response = requests.get(uri)
+
+        if response.status_code == requests.codes.ok:
+            # See http://docs.aws.amazon.com/AlexaTopSites/latest/index.html?QUERY_QueryRequests.html  # nopep8
+            dom = parseString(response.content)
+            sites = dom.getElementsByTagName("aws:Site")
+            for site in sites:
+                url = self.get_text(
+                    site.getElementsByTagName("aws:DataUrl")[0].childNodes)
+                rank = int(self.get_text(
+                    site.getElementsByTagName("aws:Rank")[0].childNodes))
+
+                priority = begin_priority
+                if rank > 100 and rank <= 1000:
+                    priority = priority + 1
+                elif rank > 1000 and rank <= 10000:
+                    priority = priority + 2
+
+                if url in topsites:
+                    site_row = topsites[url]
+                    # Update priority is new is lower
+                    if site_row.priority > priority:
+                        site_row.priority = priority
+                        site_row.country_code = self.country_code
+                        site_row.ranking = rank
+                else:
+                    site_row = Site(url, priority, self.country_code, rank)
+                    topsites[url] = site_row
+                    session.add(site_row)
+
+            session.commit()
+
+    def build_query(self, start, count):
+        '''Build query string for request with start ranking and count'''
+        now = time.time()
+        mlsec = repr(now).split('.')[1][:3]
+        nowtime = datetime.datetime.utcnow()
+        timestamp = nowtime.strftime(ATS_DATEFORMAT).format(mlsec)
+
+        query_params = {
+            "Action": ATS_ACTION_NAME,
+            "ResponseGroup": ATS_RESPONSE_GROUP_NAME,
+            "AWSAccessKeyId": self.access_key,
+            "Timestamp": timestamp,
+            "CountryCode": self.country_code,
+            "Count": count,
+            "Start": start,
+            "SignatureVersion": "2",
+            "SignatureMethod": ATS_HASH_ALGORITHM
+        }
+
+        # alexa top site only accept request with ordered query parameters
+        return urlencode(collections.OrderedDict(sorted(query_params.items())))
+
```

Comment:
yup and because we drop the milliseconds. Your code becomes a lot easier in one shot.

```python

    ATS_DATEFORMAT = '%Y-%m-%dT%H:%M:%SZ'

    def build_query(self, start, count):
        timestamp = datetime.datetime.utcnow().strftime('ATS_DATEFORMAT')
        # Alexa top site only accept request with ordered query parameters
        # Keep the order!
        query_params = [
            ("Action", ATS_ACTION_NAME),
            ("AWSAccessKeyId", access_key),
            ("Count", count),
            ("CountryCode", country_code),
            ("ResponseGroup", ATS_RESPONSE_GROUP_NAME),
            ("SignatureMethod", ATS_HASH_ALGORITHM),
            ("SignatureVersion", "2"),
            ("Start", start),
            ("Timestamp", timestamp)]
        return urlencode(query_params)
```